rust/relay/ebpf-turn-router/src/move_headers.rs

```diff
@@ -11,77 +12,201 @@ use network_types::{
+    remove_channel_data_header::<{ Ipv4Hdr::LEN }>(ctx)
 }
 
 #[inline(always)]
-pub fn add_channel_data_header_ipv4(ctx: &XdpContext, mut header: CdHdr) -> Result<(), Error> {
-    move_headers::<{ -(CdHdr::LEN as i32) }, { Ipv4Hdr::LEN }>(ctx)?;
-    let offset = (EthHdr::LEN + Ipv4Hdr::LEN + UdpHdr::LEN) as u32;
+pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
+    remove_channel_data_header::<{ Ipv6Hdr::LEN }>(ctx)
+}
+
+/// Adds a channel data header to the UDP payload by extending the packet at the tail
+/// and shifting the UDP payload backwards in 32-byte chunks (plus a small tail).
+#[inline(always)]
+fn add_channel_data_header<const IP_HEADER_LEN: usize>(
+    ctx: &XdpContext,
+    header: CdHdr,
+) -> Result<(), Error> {
+    let data_start = ctx.data();
+    let data_end = ctx.data_end();
+    let payload_offset = EthHdr::LEN + IP_HEADER_LEN + UdpHdr::LEN;
+    let payload_len = data_end - data_start - payload_offset;
+
+    // Sanity check to help verifier
+    if data_start + payload_offset > data_end {
+        return Err(Error::PacketTooLong);
+    }
 
-    let header_ptr = &mut header as *mut _ as *mut c_void;
-    let header_len = core::mem::size_of_val(&header) as u32;
+    // Constant loop bound to help verifier
+    let max_iterations = (MTU - payload_offset) / CHUNK_SIZE + 1;
 
-    let ret = unsafe { bpf_xdp_store_bytes(ctx.ctx, offset, header_ptr, header_len) };
+    // Extend the packet to make room for the header
+    let ret = unsafe { bpf_xdp_adjust_tail(ctx.ctx, CdHdr::LEN as i32) };
     if ret < 0 {
-        return Err(Error::XdpStoreBytesFailed(ret));
+        return Err(Error::XdpAdjustTailFailed(ret));
     }
 
-    Ok(())
+    // Move full 32-byte blocks from end to start
+    let mut remaining = payload_len;
+    let mut loop_count = 0;
+
+    while remaining >= CHUNK_SIZE && loop_count <= max_iterations {
+        let src = payload_offset + remaining - CHUNK_SIZE;
+        let dst = src + CdHdr::LEN;
+
+        copy_n::<32>(ctx, src, dst)?;
+
+        remaining -= CHUNK_SIZE;
+        loop_count += 1;
+    }
+
+    // Tail: move remaining 1..31 bytes using fixed constants {16,8,4,2,1}
+    if remaining > 0 && loop_count < max_iterations {
+        let mut src = payload_offset;
+        let mut dst = src + CdHdr::LEN;
+
+        tail_copy(ctx, &mut src, &mut dst, remaining)?;
+    }
+
+    // Write the channel data header at the start of the UDP payload
+    write_header_at_offset(ctx, &header, payload_offset as u32)
 }
 
+/// Removes a channel data header from the UDP payload by shifting the payload forward
+/// in 32-byte chunks (plus a small tail) and then shrinking the packet at the tail.
 #[inline(always)]
-pub fn remove_channel_data_header_ipv6(ctx: &XdpContext) -> Result<(), Error> {
-    move_headers::<{ CdHdr::LEN as i32 }, { Ipv6Hdr::LEN }>(ctx)
+fn remove_channel_data_header<const IP_HEADER_LEN: usize>(ctx: &XdpContext) -> Result<(), Error> {
```

Comment:
We shouldn't need to adjust this code path? When we remove the channel data header, we can just overwrite it.

Reply:
Hmm I was thinking we'd keep the same helper to move bytes whether it's forward or backwards so that we have less codepaths to test.

You're thinking we have one codepath that shifts headers backward (the current thing for `remove_channel_data_header`) and one codepath that shifts the payload backward (for 'add_channel_data_header`)?